File: source/blog-web/src/utils/comparator.ts
import type { BlogPostMeta, SeriesMeta } from '../types/post.ts'

/**
 * 日期降序比较器（最新日期排在最前）
 * 使用 Date.parse 避免高频实例化 Date 对象，零堆内存分配
 */
export function compareDateDesc(dateA?: string, dateB?: string): number {
  const timeA = dateA ? Date.parse(dateA) || 0 : 0
  const timeB = dateB ? Date.parse(dateB) || 0 : 0
  return timeB - timeA
}

/**
 * 可选数值降序比较器（数值越大越靠前；有值优先于无值）
 * 遵循确定性裁决，消除分散在各处的判空样板代码
 */
export function compareOptionalDesc(a?: number | null, b?: number | null): number {
  const hasA = a !== undefined && a !== null
  const hasB = b !== undefined && b !== null
  if (hasA && hasB) {
    return b - a
  }
  if (hasA) return -1
  if (hasB) return 1
  return 0
}

/**
 * 同一合集内部章节序号比较（降序：最新发表的章节排在前面）
 */
export function compareSameSeriesPartDesc(a: BlogPostMeta, b: BlogPostMeta): number {
  if (a.series?.slug && b.series?.slug && a.series.slug === b.series.slug) {
    return (b.series.part ?? 0) - (a.series.part ?? 0)
  }
  return 0
}

/**
 * 跨类别文章分类比较（未设显式权重时，独立专题文章排在合集之前，防止长合集垄断首页）
 */
export function compareCategory(a: BlogPostMeta, b: BlogPostMeta): number {
  const isSeriesA = !!a.series?.slug
  const isSeriesB = !!b.series?.slug
  if (isSeriesA !== isSeriesB) {
    return isSeriesA ? 1 : -1
  }
  return 0
}

/**
 * 全局文章复合比较器（工业级短路比较器链）
 * 严格遵循 6 级确定性规则：
 * 1. 发布时间 date 降序
 * 2. 文章级权重 order 降序
 * 3. 同一合集内按章节 part 降序
 * 4. 不同合集间按合集 series.order 降序
 * 5. 跨类别同日发布：独立文章排在合集之前
 * 6. 确定性稳定兜底：按 slug 字母升序
 */
export function comparePosts(a: BlogPostMeta, b: BlogPostMeta): number {
  return (
    compareDateDesc(a.date, b.date) ||
    compareOptionalDesc(a.order, b.order) ||
    compareSameSeriesPartDesc(a, b) ||
    compareOptionalDesc(a.series?.order, b.series?.order) ||
    compareCategory(a, b) ||
    a.slug.localeCompare(b.slug)
  )
}

/**
 * 专栏合集复合比较器（工业级比较器链）
 * 1. 权重 order 降序
 * 2. 最新更新时间 updatedAt 降序
 * 3. 确定性稳定兜底：按 slug 字母升序
 */
export function compareSeries(a: SeriesMeta, b: SeriesMeta): number {
  return (
    compareOptionalDesc(a.order, b.order) ||
    compareDateDesc(a.updatedAt, b.updatedAt) ||
    a.slug.localeCompare(b.slug)
  )
}

File: deploy/wechat/comparator.py
# -*- coding: utf-8 -*-
"""
comparator.py: 发布选择器文章排序比较器

对偶于博客前端 source/blog-web/src/utils/comparator.ts（Feed 降序），
本模块面向 CLI 编号选择做升序编排，并保证同系列连续。

职责边界：只做比较与排序，不扫描文件系统、不解析 Markdown。
"""
from functools import cmp_to_key
from typing import Any, Dict, List, Optional

from fields import date_key, part_key


def compare_optional_asc(a: Optional[float], b: Optional[float]) -> int:
    """数值升序；有值优先于无值（对偶于博客 compareOptionalDesc）。"""
    has_a = a is not None
    has_b = b is not None
    if has_a and has_b:
        return (a > b) - (a < b)
    if has_a:
        return -1
    if has_b:
        return 1
    return 0


def compare_same_series_part_asc(a: Dict[str, Any], b: Dict[str, Any]) -> int:
    """同一合集内按章节 part 升序（与 sort_series_chapters 规则一致）。"""
    slug_a = a.get("series_slug")
    slug_b = b.get("series_slug")
    if slug_a and slug_b and slug_a == slug_b:
        part_a = part_key(a.get("part"))
        part_b = part_key(b.get("part"))
        return (part_a > part_b) - (part_a < part_b)
    return 0


def compare_category(a: Dict[str, Any], b: Dict[str, Any]) -> int:
    """跨类别：独立文章排在合集之前（对齐博客 compareCategory）。"""
    is_series_a = bool(a.get("series_slug"))
    is_series_b = bool(b.get("series_slug"))
    if is_series_a != is_series_b:
        return 1 if is_series_a else -1
    return 0


def compare_posts_asc(a: Dict[str, Any], b: Dict[str, Any]) -> int:
    """
    发布选择器升序比较器 —— 对偶于博客前端 comparePosts（降序 Feed）。

    短路链：
    1. 发布时间 date 升序
    2. 文章级权重 order 升序
    3. 同一合集内按章节 part 升序
    4. 不同合集间按合集 series_order 升序
    5. 跨类别同日：独立文章排在合集之前
    6. 确定性兜底：slug 字母升序
    """
    date_cmp = (date_key(a.get("date")) > date_key(b.get("date"))) - (
        date_key(a.get("date")) < date_key(b.get("date"))
    )
    if date_cmp:
        return date_cmp

    order_cmp = compare_optional_asc(a.get("order"), b.get("order"))
    if order_cmp:
        return order_cmp

    part_cmp = compare_same_series_part_asc(a, b)
    if part_cmp:
        return part_cmp

    series_order_cmp = compare_optional_asc(a.get("series_order"), b.get("series_order"))
    if series_order_cmp:
        return series_order_cmp

    category_cmp = compare_category(a, b)
    if category_cmp:
        return category_cmp

    slug_a = a.get("slug") or ""
    slug_b = b.get("slug") or ""
    return (slug_a > slug_b) - (slug_a < slug_b)


def sort_series_chapters(chapters: List[Dict[str, Any]]) -> None:
    """专栏内按 part 升序，再按日期升序（对齐博客 loader 章节 TOC）。"""
    chapters.sort(
        key=lambda p: (part_key(p.get("part")), date_key(p.get("date")), p.get("slug") or "")
    )


def sort_posts_for_picker(posts: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """
    CLI 发布选择器排序：
    - 同系列聚合为连续块，块内 part 升序，避免连载章节被打散
    - 系列块锚点取最新章节日期，近期连载靠近列表尾部
    - 独立文章与系列块再按 compare_posts_asc 的时间升序规则交织
    """
    series_groups: Dict[str, List[Dict[str, Any]]] = {}
    standalones: List[Dict[str, Any]] = []

    for post in posts:
        series_slug = post.get("series_slug")
        if series_slug:
            series_groups.setdefault(series_slug, []).append(post)
        else:
            standalones.append(post)

    blocks: List[Dict[str, Any]] = []

    for chapters in series_groups.values():
        sort_series_chapters(chapters)
        anchor = max((date_key(p.get("date")) for p in chapters), default="1970-01-01")
        representative = {
            "date": anchor,
            "order": None,
            "part": None,
            "series_slug": chapters[0].get("series_slug"),
            "series_order": chapters[0].get("series_order"),
            "slug": chapters[0].get("series_slug") or chapters[0].get("slug") or "",
        }
        blocks.append({"kind": "series", "rep": representative, "posts": chapters})

    for post in standalones:
        blocks.append({"kind": "single", "rep": post, "posts": [post]})

    blocks.sort(key=cmp_to_key(lambda x, y: compare_posts_asc(x["rep"], y["rep"])))

    ordered: List[Dict[str, Any]] = []
    for block in blocks:
        ordered.extend(block["posts"])
    return ordered
